library/core/src/str/converts.rs

```diff
@@ -170,7 +171,13 @@ pub const fn from_utf8_mut(v: &mut [u8]) -> Result<&mut str, Utf8Error> {
 pub const unsafe fn from_utf8_unchecked(v: &[u8]) -> &str {
     // SAFETY: the caller must guarantee that the bytes `v` are valid UTF-8.
     // Also relies on `&str` and `&[u8]` having the same layout.
-    unsafe { mem::transmute(v) }
+    unsafe {
+        assert_unsafe_precondition!(
+            "str::from_utf8_unchecked requires that the input be valid UTF-8",
+            (v: &[u8]) => run_utf8_validation(v).is_ok()
+        );
+        mem::transmute(v)
+    }
```

Comment:
This check is likely to be both slow and never optimized out by LLVM, the condition is just too complicated. I'm not a library maintainer, but I'd rather defer this check in particular until we come up with another way to request it which is independent of `debug-assertions`.